sympy/core/function.py

```diff
@@ -1773,12 +1778,19 @@ def _eval_subs(self, old, new):
             return self
 
     def _eval_derivative(self, s):
-        if s not in self.free_symbols:
-            return S.Zero
-        return Add((Subs(self.expr.diff(s), self.variables, self.point).doit()
-            if s not in self.variables else S.Zero),
-            *[p.diff(s) * Subs(self.expr.diff(v), self.variables,
-            self.point).doit() for v, p in zip(self.variables, self.point)])
+        # Apply the chain rule of the derivative on the substitution variables:
+        val = Add.fromiter(p.diff(s) * Subs(self.expr.diff(v), self.variables, self.point).doit() for v, p in zip(self.variables, self.point))
+
+        # Check if there are free symbols in `self.expr`:
+        # First get the `expr_free_symbols`, which returns the free symbols
+        # that are directly contained in an expression node (i.e. stop
+        # searching if the node isn't an expression). At this point turn the
+        # expressions into `free_symbols` and check if there are common free
+        # symbols in `self.expr` and the deriving factor.
```

Comment:
What is the motivation for this?

Reply:
`A[i].free_symbols` returns `A[i]`, `A`, `i`. In `Subs` the free symbols are the symbols that are not substituted, that is the free symbols of the expression minus the variables plus the free symbols in point. If we expand `A[i]` into `A[i]`, `A`, `i`, then there could be cancellations of some of them for terms like `A[j]`. So we get the `expr_free_symbols`, which only returns `A[i]` for `A[i]`, after that step we expand into the customary free symbols.

Reply:
I think I see the need for this (also c.f. this mailing list thread https://groups.google.com/forum/#!msg/sympy/RvPrX1BhyC8/fCyJFn0RCAAJ). But it does feel a little bit like it duplicates the _diff_wrt API to some degree. Also there is probably a better name that can be used than expr_free_symbols. 

I'll think on this.

Reply:
Well, you think they `expr_free_symbols` should stop the recursion whenever it meets `_diff_wrt == True`? I don't think it's a duplicate, it's just a detection of atomic expressions.